File: src/api/RiskAnalytics.Api.Business/Services/UtilizationService.cs
using System.Security.Claims;
using System.Text;
using Microsoft.Extensions.Caching.Memory;
using RiskAnalytics.Api.Business.Authorization;
using RiskAnalytics.Api.Business.Mappers.Interfaces;
using RiskAnalytics.Api.Business.Services.Interfaces;
using RiskAnalytics.Api.Common;
using RiskAnalytics.Api.Model;
using RiskAnalytics.Api.Repository;
using RiskAnalytics.Api.Repository.Interfaces;
using RiskAnalytics.Api.Repository.Models;

namespace RiskAnalytics.Api.Business.Services;

public class UtilizationService : IUtilizationService
{
    private readonly IUtilizationRepository utilizationRepository;
    private readonly IPortfolioAuthorizationService portfolioAuthorizationService;
    private readonly IMemoryCache memoryCache;
    private readonly IMonthlyUtilizationMapper monthlyUtilizationMapper;
    private readonly IGroupCountQueryRepository groupCountQueryRepository;

    public UtilizationService(
        IPortfolioAuthorizationService portfolioAuthorizationService,
        IUtilizationRepository utilizationRepository,
        IGroupCountQueryRepository groupCountQueryRepository,
        IMemoryCache memoryCache,
        IMonthlyUtilizationMapper monthlyUtilizationMapper)
    {
        this.portfolioAuthorizationService = portfolioAuthorizationService;
        this.utilizationRepository = utilizationRepository;
        this.memoryCache = memoryCache;
        this.monthlyUtilizationMapper = monthlyUtilizationMapper;
        this.groupCountQueryRepository = groupCountQueryRepository;
    }

    public async Task<IEnumerable<IEnumerable<MonthlyUtilization>>> GetMonthlyUtilization(
        int? portfolioId,
        ClaimsPrincipal user,
        MonthlyUtilizationGroup? groupBy,
        IReadOnlyCollection<int>? groupByFilterIds,
        int? operatorId,
        int? lessorId,
        bool includeBaseline,
        bool isEmissions,
        bool isHoursAndCycle)
    {
        if (portfolioId != null)
        {
            await portfolioAuthorizationService.ValidateAccessToPortfolioOrThrow(portfolioId.Value, user.Auth0Id());
        }

        var includeEmissions = user.Claims.HasPortfoliosClaim(PortfoliosClaim.EmissionsAddOn);
        if (portfolioId != null || (groupBy != null && groupBy != MonthlyUtilizationGroup.MarketClass) || operatorId != null || lessorId != null)
        {

            return await BuildMonthlyUtilization();
        }

        var cacheKey = BuildCacheKey();
        IEnumerable<IEnumerable<MonthlyUtilization>> cachedMonthlyUtilization;

        var cached = memoryCache.TryGetValue(cacheKey, out cachedMonthlyUtilization);

        if (cached && cachedMonthlyUtilization != null)
        {
            return cachedMonthlyUtilization;
        }

        var monthlyUtilization = await BuildMonthlyUtilization();

        memoryCache.Set(cacheKey, monthlyUtilization, TimeSpan.FromMinutes(CacheSettings.MonthlyUtilizationCachePeriodInMinutes));

        return monthlyUtilization;

        async Task<IList<IEnumerable<MonthlyUtilization>>> BuildMonthlyUtilization()
        {

            var monthlyUtilizationsTask = utilizationRepository.GetMonthlyUtilization(
                portfolioId,
                groupBy,
                groupByFilterIds,
                operatorId,
                lessorId,
                includeEmissions,
                includeBaseline,
                isEmissions,
                isHoursAndCycle);

            var monthlyUtilizations = (await monthlyUtilizationsTask).ToList();

            monthlyUtilization = await GetGroupCounts(portfolioId,
                groupBy,
                groupByFilterIds,
                operatorId,
                lessorId,
                monthlyUtilizations,
                includeBaseline);

            return monthlyUtilizationMapper.Map(monthlyUtilizations);
        }

        string BuildCacheKey()
        {
            var cacheKeyBuilder = new StringBuilder();

            cacheKeyBuilder.Append(CacheSettings.MonthlyUtilizationGlobalBenchmarkResultsCachePrefix);
            cacheKeyBuilder.Append($"_IncludeBaseline={includeBaseline}");
            cacheKeyBuilder.Append($"_MarketClassIds={(groupByFilterIds == null ? "All" : string.Join(',', groupByFilterIds.Order()))}");
            cacheKeyBuilder.Append($"_IncludeEmissions={includeEmissions}");
            cacheKeyBuilder.Append($"_IsEmissions={isEmissions}");
            cacheKeyBuilder.Append($"_isHoursAndCycle={isHoursAndCycle}");

            return cacheKeyBuilder.ToString();
        }
    }
    public async Task<IList<IEnumerable<MonthlyUtilization>>> GetGroupCounts(
        int? portfolioId,
        MonthlyUtilizationGroup? groupBy,
        IReadOnlyCollection<int>? groupByFilterIds,
        int? operatorId,
        int? lessorId,
        IList<IEnumerable<MonthlyUtilization>> monthlyUtilizations,
        bool includeBaseline)
    {
        // Pre-process the monthlyUtilizations into a dictionary for faster lookups
        var lookup = monthlyUtilizations
            .SelectMany(i => i)
            .GroupBy(x => new { x.GroupId, x.Year, x.Month })
            .ToDictionary(g => g.Key, g => g.First());

        var monthlyUtilizationGroupCounts = await groupCountQueryRepository.GetGroupCounts(
            portfolioId,
            groupBy,
            groupByFilterIds,
            operatorId,
            lessorId,
            includeBaseline);

        foreach (var groupCount in monthlyUtilizationGroupCounts)
        {
            var key = new { groupCount.GroupId, groupCount.Year, groupCount.Month };
            if (lookup.TryGetValue(key, out var matchingUtilization))
            {
                matchingUtilization.NumberOfAircraftInGroup = groupCount.NumberOfAircraftInGroup;
            }
        }

        return monthlyUtilizations;
    }

    public async Task<IEnumerable<MSNUtilizationPerAircraft>> GetMonthlyUtilizationPerAircraft(
        int portfolioId,
        int endMonthIndex,
        int startMonthIndex,
        int endYear,
        int startYear,
        bool isEmissions,
        ClaimsPrincipal user,
        int? operatorId,
        int? lessorId,
        MonthlyUtilizationGroup? groupBy,
        IReadOnlyCollection<int>? groupByFilterIds)
    {
        var cacheKey = this.BuildCacheKey(portfolioId, endMonthIndex, startMonthIndex, endYear, startYear, isEmissions, user, operatorId, lessorId, groupBy, groupByFilterIds);
        IEnumerable<MSNUtilizationPerAircraft> cachedMonthlyUtilizationPerAircraft;

        var cached = memoryCache.TryGetValue(cacheKey, out cachedMonthlyUtilizationPerAircraft);

        if (cachedMonthlyUtilizationPerAircraft != null)
        {
            return cachedMonthlyUtilizationPerAircraft;
        }
        var monthlyUtilizationsPerAircraft = await utilizationRepository.GetMonthlyUtilizationPerAircraft(portfolioId, endMonthIndex, startMonthIndex, endYear, startYear, isEmissions, operatorId, lessorId, groupBy, groupByFilterIds);
        memoryCache.Set(cacheKey, monthlyUtilizationsPerAircraft, TimeSpan.FromMinutes(CacheSettings.MonthlyUtilizationCachePeriodInMinutes));
        return monthlyUtilizationsPerAircraft;
    }

    public async Task<UtilizationGroupOptionsModel> GetGroupOptions(
        int? portfolioId,
        int? operatorId,
        int? lessorId,
        string userId)
    {
        if (portfolioId != null)
        {
            await portfolioAuthorizationService.ValidateAccessToPortfolioOrThrow(portfolioId.Value, userId);
        }

        if (portfolioId != null || operatorId != null || lessorId != null)
        {
            return await GetGroupOptionsFromRepository();
        }

        const string cacheKey = CacheSettings.MonthlyUtilizationGlobalBenchmarkGroupOptionsCacheKey;

        if (memoryCache.TryGetValue(cacheKey, out UtilizationGroupOptionsModel? utilizationGroupOptionsModel) && utilizationGroupOptionsModel != null)
        {
            return utilizationGroupOptionsModel;
        }

        var groupOptionsModel = await GetGroupOptionsFromRepository();

        memoryCache.Set(cacheKey, groupOptionsModel, TimeSpan.FromMinutes(CacheSettings.MonthlyUtilizationCachePeriodInMinutes));

        return groupOptionsModel;

        async Task<UtilizationGroupOptionsModel> GetGroupOptionsFromRepository()
        {
            var groupOptions = (await utilizationRepository.GetGroupOptions(operatorId, portfolioId, lessorId)).ToList();

            return new UtilizationGroupOptionsModel
            {
                AircraftMarketClasses = groupOptions.Where(x => x.Type == MonthlyUtilizationGroup.MarketClass.ToString()).Select(i => new IdNamePairModel(i.Id, i.Name)),
                AircraftFamilies = groupOptions.Where(x => x.Type == MonthlyUtilizationGroup.AircraftFamily.ToString()).Select(i => new IdNamePairModel(i.Id, i.Name)),
                AircraftTypes = groupOptions.Where(x => x.Type == MonthlyUtilizationGroup.AircraftType.ToString()).Select(i => new IdNamePairModel(i.Id, i.Name)),
                AircraftSeries = groupOptions.Where(x => x.Type == MonthlyUtilizationGroup.AircraftSeries.ToString()).Select(i => new IdNamePairModel(i.Id, i.Name)),
                AircraftSerialNumbers = groupOptions.Where(x => x.Type == MonthlyUtilizationGroup.AircraftSerialNumber.ToString()).Select(i => new IdNamePairModel(i.Id, i.Name))
            };
        }
    }

    public async Task<IEnumerable<IdNamePairModel>> GetOperators(
        int? portfolioId,
        int? lessorId,
        MonthlyUtilizationGroup? groupBy,
        IReadOnlyCollection<int>? groupByFilterIds,
        string userId)
    {
        if (portfolioId != null)
        {
            await portfolioAuthorizationService.ValidateAccessToPortfolioOrThrow(portfolioId.Value, userId);
        }

        if (portfolioId != null || groupBy != null)
        {
            return await GetOperatorsFromRepository();
        }

        if (memoryCache.TryGetValue(CacheSettings.MonthlyUtilizationGlobalBenchmarkOperatorsCacheKey, out IEnumerable<IdNamePairModel>? cachedOperators) && cachedOperators != null)
        {
            return cachedOperators;
        }

        var operators = await GetOperatorsFromRepository();

        memoryCache.Set(CacheSettings.MonthlyUtilizationGlobalBenchmarkOperatorsCacheKey, operators, TimeSpan.FromMinutes(CacheSettings.MonthlyUtilizationCachePeriodInMinutes));

        return operators;

        async Task<IList<IdNamePairModel>> GetOperatorsFromRepository()
        {
            return (await utilizationRepository.GetOperators(portfolioId, lessorId, groupBy, groupByFilterIds)).ToList();
        }
    }

    public async Task<IEnumerable<IdNamePairModel>> GetLessors(
        int? portfolioId,
        int? operatorId,
        MonthlyUtilizationGroup? groupBy,
        IReadOnlyCollection<int>? groupByFilterIds,
        string userId)
    {
        if (portfolioId != null)
        {
            await portfolioAuthorizationService.ValidateAccessToPortfolioOrThrow(portfolioId.Value, userId);
        }

        if (portfolioId != null || groupBy != null || operatorId != null)
        {
            return await GetLessorsFromRepository();
        }

        if (memoryCache.TryGetValue(CacheSettings.MonthlyUtilizationGlobalBenchmarkLessorsCacheKey, out IEnumerable<IdNamePairModel>? cachedOperators) && cachedOperators != null)
        {
            return cachedOperators;
        }

        var operators = await GetLessorsFromRepository();

        memoryCache.Set(CacheSettings.MonthlyUtilizationGlobalBenchmarkLessorsCacheKey, operators, TimeSpan.FromMinutes(CacheSettings.MonthlyUtilizationCachePeriodInMinutes));

        return operators;

        async Task<IList<IdNamePairModel>> GetLessorsFromRepository()
        {
            return (await utilizationRepository.GetLessors(portfolioId, operatorId, groupBy, groupByFilterIds)).ToList();
        }
    }

    private string BuildCacheKey(
        int portfolioId,
        int endMonthIndex,
        int startMonthIndex,
        int endYear,
        int startYear,
        bool isEmissions,
        ClaimsPrincipal user,
        int? operatorId,
        int? lessorId,
        MonthlyUtilizationGroup? groupBy,
        IReadOnlyCollection<int>? groupByFilterIds)
    {
        var cacheKeyBuilder = new StringBuilder();
        cacheKeyBuilder.Append(CacheSettings.MonthlyUtilizationPerAircraftCachePrefix);
        cacheKeyBuilder.Append($"_PortfolioId={portfolioId}");
        cacheKeyBuilder.Append($"_EndMonthIndex={endMonthIndex}");
        cacheKeyBuilder.Append($"_StartMonthIndex={startMonthIndex}");
        cacheKeyBuilder.Append($"_EndYear={endYear}");
        cacheKeyBuilder.Append($"_StartYear={startYear}");
        if (operatorId != null)
            cacheKeyBuilder.Append($"_OperatorId={operatorId}");
        if (lessorId != null)
            cacheKeyBuilder.Append($"_LessorId={lessorId}");
        if (groupBy != null)
            cacheKeyBuilder.Append($"_GroupBy={groupBy}");
        if (groupByFilterIds?.Count > 0)
            cacheKeyBuilder.Append($"_GroupByFilterIds={string.Join(',', groupByFilterIds)}");
        cacheKeyBuilder.Append($"_IsEmissions={isEmissions}");

        return cacheKeyBuilder.ToString();
    }

}

File: src/frontend/src/app/modules/shared/services/utilization.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { MonthlyUtilization } from '../../compare-portfolio/models/monthly-utilization';
import {
  OperatorOption,
  LessorOption
} from '../../compare-portfolio/components/portfolio-detail-options/portfolio-detail-options.component';
import { GroupOptions } from '../../compare-portfolio/models/group-options';
import { MSNUtilizationPerAircraft } from '../../compare-portfolio/models/monthly-utilization-per-aircraft';

@Injectable({
  providedIn: 'root'
})
export class UtilizationService {
  constructor(private readonly httpClient: HttpClient) {}

  getMonthlyUtilization(
    portfolioId: number | undefined,
    includeBaseline: boolean,
    isEmissions: boolean,
    isHoursAndCycle: boolean,
    groupBy?: string,
    groupByFilterIds?: number[],
    operatorId?: number,
    lessorId?: number
  ): Observable<MonthlyUtilization[][]> {
    let params = new HttpParams();

    if (portfolioId) {
      params = params.append('portfolioId', portfolioId);
    }

    if (groupBy) {
      params = params.append('groupBy', groupBy);
    }

    groupByFilterIds?.forEach((groupByFilterId) => {
      params = params.append('groupByFilterIds', groupByFilterId);
    });

    if (operatorId) {
      params = params.append('operatorId', operatorId);
    }

    if (lessorId) {
      params = params.append('lessorId', lessorId);
    }

    params = params.append('includeBaseline', includeBaseline);
    params = params.append('isEmissions', isEmissions);
    params = params.append('isHoursAndCycle', isHoursAndCycle);

    return this.httpClient.get<MonthlyUtilization[][]>(`/api/utilization`, { params });
  }

  getMonthlyUtilizationPerAircraft(
    portfolioId: number | undefined,
    endMonthIndex: number,
    startMonthIndex: number,
    endYear: number,
    startYear: number,
    isEmissions: boolean,
    operatorId?: number,
    lessorId?: number,
    groupBy?: string,
    groupByFilterIds?: number[]
  ): Observable<MSNUtilizationPerAircraft[]> {
    let params = new HttpParams();
    endMonthIndex = endMonthIndex + 1;
    startMonthIndex = startMonthIndex + 1;

    const paramValues = {
      portfolioId,
      endMonthIndex,
      startMonthIndex,
      endYear,
      startYear,
      isEmissions,
      operatorId,
      lessorId,
      groupBy
    };

    params = this.appendParams(params, paramValues);

    groupByFilterIds?.forEach((groupByFilterId) => {
      params = params.append('groupByFilterIds', groupByFilterId);
    });

    return this.httpClient.get<MSNUtilizationPerAircraft[]>(`/api/utilization/aircraft/monthly`, { params });
  }

  appendParams(params: HttpParams, paramValues: { [key: string]: any }): HttpParams {
    Object.keys(paramValues).forEach((key) => {
      if (paramValues[key]) {
        params = params.append(key, paramValues[key]);
      }
    });
    return params;
  }

  getOperators(portfolioId?: number, groupBy?: string, lessorId?: number, groupByFilterIds?: number[]): Observable<OperatorOption[]> {
    let params = new HttpParams();

    if (groupBy) {
      params = params.append('groupBy', groupBy);
    }

    groupByFilterIds?.forEach((groupByFilterId) => {
      params = params.append('groupByFilterIds', groupByFilterId);
    });

    if (portfolioId) {
      params = params.append('portfolioId', portfolioId);
    }

    if (lessorId) {
      params = params.append('lessorId', lessorId);
    }

    return this.httpClient.get<OperatorOption[]>(`/api/utilization/operators`, { params });
  }

  getLessors(portfolioId?: number, groupBy?: string, operatorId?: number, groupByFilterIds?: number[]): Observable<LessorOption[]> {
    let params = new HttpParams();

    if (groupBy) {
      params = params.append('groupBy', groupBy);
    }

    groupByFilterIds?.forEach((groupByFilterId) => {
      params = params.append('groupByFilterIds', groupByFilterId);
    });

    if (portfolioId) {
      params = params.append('portfolioId', portfolioId);
    }

    if (operatorId) {
      params = params.append('operatorId', operatorId);
    }

    return this.httpClient.get<LessorOption[]>(`/api/utilization/lessors`, { params });
  }

  getGroupOptions(operatorId?: number, portfolioId?: number, lessorId?: number): Observable<GroupOptions> {
    let params = new HttpParams();

    if (operatorId) {
      params = params.append('operatorId', operatorId);
    }

    if (portfolioId) {
      params = params.append('portfolioId', portfolioId);
    }

    if (lessorId) {
      params = params.append('lessorId', lessorId);
    }

    return this.httpClient.get<GroupOptions>(`/api/utilization/groups`, { params });
  }
}
